### PR TITLE
rework append_if_no_line to use file provider subresource

### PR DESCRIPTION
**Pull request by lamont-granquist**
there's a ton of the guts of the file provider missing if you don't do this.

as a concrete example here what this is doing is opening a file and then just appending to it, without going through writing it out to a tempfile and moving it into place.

you're missing all of this machinery:

https://stackoverflow.com/a/4174125/506908

which will have horrible edge conditions in situations when you run out of disk space trying to do the append, etc.

also added some tests around files-with-no-terminating-carriage-return which fail on master (badly, actually)...

```
     ×  File /tmp/file_without_linereturn content should eql "no carriage return line\nSHOULD GO ON ITS OWN LINE\n"

     expected: "no carriage return line\nSHOULD GO ON ITS OWN LINE\n"
          got: "no carriage return lineSHOULD GO ON ITS OWN LINE\nno carriage return line\n"
```

---
This change is [Reviewable](https://reviewable.io/reviews/sous-chefs/line-cookbook/74)

### CODE REVIEW

**Comment by kitchen-porter** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          2 Errors
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add labels to this Pull Request</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please include a <a href="https://github.com/sous-chefs/line-cookbook/blob/master/CHANGELOG.md">CHANGELOG</a> entry.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This Pull Request is probably missing tests.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by lamont-granquist** [comment]
one more bit incoming

**Comment by lamont-granquist** [comment]
@damacus okay shippable

File: resources/append_if_no_line.rb [hunk 1]
```diff
@@ -4,36 +4,18 @@
...
-    begin
-      f = ::File.open(new_resource.path, 'w')
-      converge_by "Updating file #{new_resource.path}" do
-        f.puts new_resource.line
-      end
-    ensure
-      f.close
-    end
+  file new_resource.path do
+    content((current + [new_resource.line + "\n"]).join)
```
Comment (MarkGibbons) [inline resources/append_if_no_line.rb:18]: Whole file in memory vs line by line, some chance putting the file in memory will break for someone.

**Comment by damacus** [comment]
This is awesome thank you.

If I had given myself some headspace Im pretty sure I would have just nicked the code from another project I had (oh the irony) that I wrote. 

I'll have a proper look over this, this weekend and merge both pull requests.

File: resources/append_if_no_line.rb (hunk 1 above)
Comment (lamont-granquist) [reply resources/append_if_no_line.rb:18]: The bugs that are fixed here are likely more concrete though

It does highlight that the file resources content property is inherently slurpy.  I can probably fix that directly in this cookbook but it'll get ugly.

**Review by damacus — APPROVED** [review]

File: resources/append_if_no_line.rb (hunk 1 above)
Comment (damacus) [reply resources/append_if_no_line.rb:18]: rewriting very large files that will fill up the ram seems not ideal anyway. And probably outside the main use cases of line imho.

**Comment by lamont-granquist** [comment]
^ i came up with a non-slurpy version of append_if_no_line but its pretty ugly -- without bug reports i don't think it'd be worth it.

File: resources/append_if_no_line.rb (hunk 1 above)
Comment (MarkGibbons) [reply resources/append_if_no_line.rb:18]: I was thinking more in terms of version 2.0.0 vs a minor version bump and fixing bug reports if we got any.  I've thought of rewriting this the way your did for years but never did because I figured someone would object to slurping the file.
